Complete TPTP support

eugenk opened the pull request:
This adds the TPTP logic with prover support. It supports the whole TPTP syntax, including CNF, FOF, TFF and THF.

tillmo commented:
TPTP only allows imports, so inclusion signature morphisms suffice. Of course, at a later stage, full signature morphisms (for use with DOL constructs) would be nice.

mcodescu commented on TPTP/Sublogic.hs:
```diff
@@ -0,0 +1,135 @@
+sublogicOfUnit :: Sublogic -> () -> Sublogic
+sublogicOfUnit sublogic () = bottom
+
+sublogicOfBaiscSpec :: Sublogic -> BASIC_SPEC -> Sublogic
```
Would rename this to `sublogicOfBasicSpec`

mcodescu commented:
I have taken a thorough look at the code, it's very clean. In some files there is commented-out code in large chunks, which could be removed. I have also translated some CASL specifications to TPTP and looked at the resulting theory. I wanted to test all the provers that were added via TPTP. After compilation, I don't see CVC4, leo-2 (but I have leo available as an option and I get an error message that the prover is missing when trying to select it), geo3 and satallax. I would guess I need to install them myself, since they don't come with hets (at least didn't come when I installed the package). It would be useful to maintain a page saying "if you don't have prover X from the hets package, install it from here", especially if we get new Hets developers.

tillmo commented:
@mcodescu, have you installed `hets-provers`?

eugenk commented:
There are no Ubuntu-packages for these provers, yet. I have compiled them myself from the sources at www.cs.miami.edu/~tptp/CASC/J8/SystemsSources/

eugenk commented:
> I found it a bit strange to have the type Morphism defined in a different file than Morphism.hs

done with 7f19b49cb

> the file Logic.hs should be named Logic_TPTP.hs

done with 8d9547cfe

mcodescu commented:
@tillmo I have now and that has added some things, but for example leo is still missing. I would say this is not essential for this pull request and we should discuss the issue elsewhere.

tillmo commented:
Indeed, I have installed leo manually from http://page.mi.fu-berlin.de/cbenzmueller/leo/ This should be packaged...

jelmd commented:
If in doubts, you may also check http://pkg.cs.ovgu.de/LNF/linux/ubuntu/15.04/. `15.04` or anything else, but since ontohub currently runs on 15.04, chances are higher, that add. packages were made for this release.

One can install the *.deb with `dpkg -i ${file}.deb`. They are usually cross-release compatible.

If you wanna know the packaged version of the ${file}.deb you may check  http://pkg.cs.ovgu.de/LNF/linux/ubuntu/15.04/LNF${file}/pkginfo or download the file and do a `dpkg-deb -R ${file}.deb tmp` and than check `tmp/DEBIAN/control`.

eugenk commented:
> In some files there is commented-out code in large chunks, which could be removed.

The comments there are intentional because the TPTP has a very large grammar that is still being developed. Some of the commented out code might be useful in future TPTP language versions.

mcodescu commented:
I have tested this branch on a theory I obtained by translating on the branch CASL2TPTP from CASL to TPTP (test_cond from Till's examples on that branch). This works fine, except I had to remove the sublogic from the logic declaration in the generated theory (only `logic TPTP` instead of `logic TPTP.FOF`). I think the convention is that the result of translating a theory in Hets should be a valid input for Hets again. If so, then this needs to be fixed. 

I will test this branch also on some TPTP files written by other people, and report on the results.

tillmo commented:
actually, Hets should be able to read in a file with `logic TPTP.FOF`.

mcodescu commented:
This is what I get on the branch complete_tptp_support: 
```
./hets -g test_cond.het 
Analyzing file test_cond.het as library test_cond
logic TPTP.FOF
/home/mcodescu/repos/Hets2/test_cond.het:1.1: *** Error:
unknown sublogic of logic TPTP: FOF

```

tillmo commented:
I think that this behaviour may be caused by the missing implementation of `all_sublogics`.

mcodescu commented:
With the file here: http://www.cs.miami.edu/~tptp/cgi-bin/SeeTPTP?Category=Problems&Domain=ALG&File=ALG001^5.p (you need to download it, it does not work by calling hets on the URL), I get a theory with an open proof goal. Then when calling Prove on the node, I get a list of available provers (including Leo-II, Geo-III, CVC4). In most of the cases, after selecting a prover, I get a new window with the title (...- Isabelle prover), although the prover selected was not Isabelle. I get this for SPASS too, and I know SPASS is available on my machine, so this should not be the reason. This would not disturb me so much, but proving itself does not work: after selecting the goal and clicking the Prove button, I get a message "Could not determine the SZS status". It's hard for me to judge if this is expected. 

Otherwise, I have tested imports and they work as expected.

tillmo commented:
I can confirm this behaviour.

eugenk commented:
I'm on it. Where can I find the test_cond.het?

mcodescu commented:
I made it available here: https://ontohub.org/sandbox/test_cond.het

eugenk commented:
I rebased this branch on master (so Stack is supposed to be used) and added the commit "Implement all_sublogics."

This allows to read the test_cond.het.

eugenk commented:
It's odd that when I click on `Prove` in the GUI menu of the DGNode of `ALG001^5.p`, then SPASS, EProver and other first order provers are suggested (at least on my machine) for this higher order problem even though there is no translation to first order. Is this normal?

mcodescu commented:
Till, but SPASS was shown, it just sends me to a window with Isabelle prover in the title.

mcodescu commented:
These are the provers available on my machine: CVC4, Darwin, EProver, Geo-III, Isabelle, Leo-II, SPASS, Satallax, Vampire.

eugenk commented:
I may be out on a limb here, but I think there's something different wrong with proving on TPTP in Hets: It shows provers as available, even though they aren't. I don't have a `geo3` executable in my PATH, but Hets shows me Geo-III when I click on `Prove` on a node. This should not happen.

There should be a check whether or not the executable is available somewhere in the code, but I did not implement one in this branch and thought this was already done somewhere else in Hets.

eugenk commented:
Can you point me to what I need to do to implement this check?

tillmo commented:
`mkUsableProver` in `Logic.Prover` checks for the prover binary and stores the result in `proverUsable`. This works for me e.g. for `leo`.

mcodescu commented:
I have also tested the behavior on a first-order problem. With the file here http://www.cs.miami.edu/~tptp/cgi-bin/SeeTPTP?Category=Problems&Domain=COL&File=COL001-1.p I have changed from negated_conjecture to conjecture and removed the negation in the sentence. SPASS tries to prove the theorem and timeouts.

eugenk commented:
I can reproduce the SZS status error with EProver on that file:
<img width="755" alt="screen shot 2017-08-16 at 09 55 13" src="https://user-images.githubusercontent.com/339751/29353073-442b79b0-8269-11e7-97ff-fd15c50185be.png">


What do you get when you click on "Show details" after attempting to prove it (with the SZS status error)?
![screen shot 2017-08-16 at 09 55 29](https://user-images.githubusercontent.com/339751/29353022-1cb04e10-8269-11e7-9425-dad3ee1ae30d.png)

eugenk commented:
I rebased on master and added a commit "Recognise negated_conjecture as a proof obligation."

eugenk commented:
@tillmo 
> `mkUsableProver` in `Logic.Prover` checks for the prover binary and stores the result in `proverUsable`. This works for me e.g. for `leo`.  This information is used in `Proofs.AbstractState` in function `getUsableProvers`. In `Proofs.InferBasic`, this function is used for selecting the provers sent to the GUI.

I used `mkAutomaticProver` in TPTP/Prover/Common.hs which in turn uses `mkUsableProver`. I assume there is something different wrong.

I debugged it a little bit and got this for the `proverUsable` field:
```
proverUsable cvc4: Just "missing binary in $PATH: cvc4"
proverUsable darwin: Nothing
proverUsable eprover: Nothing
proverUsable geo3: Just "missing binary in $PATH: geo3"
proverUsable isabelle: Nothing
proverUsable leo: Nothing
proverUsable satallax: Just "missing binary in $PATH: satallax"
proverUsable SPASS: Nothing
proverUsable vampire: Nothing
```

And the function `Proofs.InferBasic.basicInferenceNode` gets this after calling `getUsableProvers ProveGUI`
```
[(Vampire,id_TPTP.FOF : TPTP -> TPTP),
(SPASS,id_TPTP.FOF : TPTP -> TPTP),
(EProver,id_TPTP.FOF : TPTP -> TPTP),
(Darwin,id_TPTP.FOF : TPTP -> TPTP),
(Leo-II,id_TPTP.FOF : TPTP -> TPTP),
(Isabelle,id_TPTP.FOF : TPTP -> TPTP),

(Vampire,id_TPTP.FOF : TPTP -> TPTP),
(SPASS,id_TPTP.FOF : TPTP -> TPTP),
(EProver,id_TPTP.FOF : TPTP -> TPTP),
(Darwin,id_TPTP.FOF : TPTP -> TPTP),
(Leo-II,id_TPTP.FOF : TPTP -> TPTP),
(Isabelle,id_TPTP.FOF : TPTP -> TPTP)]
```

tillmo commented:
This is strange. Can you please trace the argument `knownProvers` of function `showProverGUI` in `GUI.GtkProverGUI`?

eugenk commented:
That file is not even compiled on my machine.

tillmo commented:
oh, then `GUI.ProverGUI`

mcodescu commented:
With EProver:

```
eprover: /tmp/COL001-1.p_COL001-1.p_prove_fixed_point5965166491189641421.tptp:9:(Column 24):(just read 'conjecture'): Identifier (axiom|definition|theorem|assumption|hypothesis|negated_conjecture|lemma|unknown|plain|watchlist) expected, but Identifier not terminating in a number('conjecture') read 
```

eugenk commented:
@mcodescu, I think this is because the TPTP syntax has changed since your EProver version has been implemented. At least I had this problem several times with EProver 1.9. A newer version of EProver might fix this.

eugenk commented:
@tillmo: this is the `knownProvers` argument of `GUI.ProverGUI.proverGUI` filtered for provers with TPTP as the target:

```
fromList [
("CVC4",[id_TPTP.THF : TPTP -> TPTP]),
("Darwin",[id_TPTP.THF : TPTP -> TPTP]),
("EProver",[id_TPTP.THF : TPTP -> TPTP]),
("Geo-III",[id_TPTP.THF : TPTP -> TPTP]),
("Isabelle",[id_TPTP.THF : TPTP -> TPTP]),
("Leo-II",[id_TPTP.THF : TPTP -> TPTP]),
("Satallax",[id_THF.THF : THF -> THF,id_TPTP.THF : TPTP -> TPTP]),
]
```

tillmo commented:
Hm, strange. Then please trace variable `ps` and the term `sublogicOfTh thForProof`in function `basicInferenceNode` in `Proofs.InferBasic`.
And what does `getUsableProvers ProveGUI s` in  `Proofs.AbstractState` give, if `s` is the value of the term `sublogicOfTh thForProof` above?

eugenk commented:
```
Proofs.InferBasic.basicInferenceNode

ps:
[(Vampire,id_TPTP.FOF : TPTP -> TPTP),
(SPASS,id_TPTP.FOF : TPTP -> TPTP),
(EProver,id_TPTP.FOF : TPTP -> TPTP),
(Darwin,id_TPTP.FOF : TPTP -> TPTP),
(Leo-II,id_TPTP.FOF : TPTP -> TPTP),
(Isabelle,id_TPTP.FOF : TPTP -> TPTP),
(Vampire,id_TPTP.FOF : TPTP -> TPTP),
(SPASS,id_TPTP.FOF : TPTP -> TPTP),
(EProver,id_TPTP.FOF : TPTP -> TPTP),
(Darwin,id_TPTP.FOF : TPTP -> TPTP),
(Leo-II,id_TPTP.FOF : TPTP -> TPTP),
(Isabelle,id_TPTP.FOF : TPTP -> TPTP)]

sublogicOfTh thForProof:
TPTP.FOF Proofs.InferBasic.basicInferenceNode
  [(Vampire,id_TPTP.FOF : TPTP -> TPTP),
  (SPASS,id_TPTP.FOF : TPTP -> TPTP),
  (EProver,id_TPTP.FOF : TPTP -> TPTP),
  (Darwin,id_TPTP.FOF : TPTP -> TPTP),
  (Leo-II,id_TPTP.FOF : TPTP -> TPTP),
  (Isabelle,id_TPTP.FOF : TPTP -> TPTP),
  (Vampire,id_TPTP.FOF : TPTP -> TPTP),
  (SPASS,id_TPTP.FOF : TPTP -> TPTP),
  (EProver,id_TPTP.FOF : TPTP -> TPTP),
  (Darwin,id_TPTP.FOF : TPTP -> TPTP),
  (Leo-II,id_TPTP.FOF : TPTP -> TPTP),
  (Isabelle,id_TPTP.FOF : TPTP -> TPTP)]
```

tillmo commented:
CVC4 is now missing htere - are you sure that this is all about the same proof goal?
Perhaps we should try to debug this jointly on Monday.

eugenk commented:
> are you sure that this is all about the same proof goal? 

Pretty sure. 

> Perhaps we should try to debug this jointly on Monday.

Ok

eugenk commented:
I opened #1725 for the problem with the unavailable provers.

eugenk commented:
#1724 should solve the problem with the newer TPTP syntax

tillmo commented:
@mcodescu: it seems that all problems have been solved or relegated to other tickets. Anything left?

mcodescu commented:
No, everything looks good for the moment. Maybe it would be a good idea to mirror the entire TPTP library on Ontohub, and see if everything works as expected. 

Otherwise, :+1:

eugenk commented:
Before doing that, I'd like to upgrade TPTP support to version 7.0.0 in another pull request (this pull request implements 6.4.0.11).
